lib/engine/distance_graph.rb

```diff
@@ -0,0 +1,139 @@
+# frozen_string_literal: true
+
+module Engine
+  class DistanceGraph
+    def initialize(game, **opts)
+      @game = game
+      @node_distances = {}
+      @path_distances = {}
+      @hex_distances = {}
+      @separate_node_types = opts[:separate_node_types] || false
+    end
+
+    def clear
+      @node_distances.clear
+      @path_distances.clear
+      @hex_distances.clear
+    end
+
+    def node_distances(corporation)
+      compute(corporation) unless @node_distances[corporation]
+      @node_distances[corporation]
+    end
+
+    def path_distances(corporation)
+      compute(corporation) unless @path_distances[corporation]
+      @path_distances[corporation]
+    end
+
+    def hex_distances(corporation)
+      compute(corporation) unless @hex_distances[corporation]
+      @hex_distances[corporation]
+    end
+
+    def get_token_cities(corporation)
+      tokens = []
+      @game.hexes.each do |hex|
+        hex.tile.cities.each do |city|
+          next unless city.tokened_by?(corporation)
+
+          tokens << city
+        end
+      end
+      tokens
+    end
+
+    def smaller_or_equal_distance?(a, b)
+      a&.all? { |k, v| v <= b[k] }
+    end
+
+    def merge_distance(a, b)
+      return b.dup unless a
+
+      a.map { |k, v| [k, [v, b[k]].min] }.to_h
+    end
+
+    def node_walk(
+      node,
+      distance: nil,
+      node_distances: {},
```

Comment:
why make these default? do you ever pass it whit nil?

Reply:
Fixed.